### PR TITLE
add beforeQueryHelper

### PR DESCRIPTION
**Pull request by antoniopresto**
I'll update tests, for now, is this acceptable?

I need beforeQuery to handle custom user/regions/etc rules.

### CODE REVIEW

**Comment by nodkz** [comment]
# Yep, graphql-compose-mongoose needs in this changes ;) 


## My thoughts


### 1) I think that we need to pass mongoose model to 

```diff
- const res = await beforeQueryHelper(resolveParams).exec();
+ const res = await beforeQueryHelper(resolveParams, model).exec();

```

It helps to recreate query from scratch, make additional queries, check what models are using for cases when one `beforeQuery` are used for different models.

### 2) We need to hide `exec()` in `beforeQueryHelper`.

It helps mock results or return data from other sources eg cache.

```
- const res = await beforeQueryHelper(resolveParams, model).exec();
+ const res = await beforeQueryHelper(resolveParams, model);
```

PS. Count resolver should be rewritten.

```diff
- resolveParams.query = model.find();
- return beforeQueryHelper(resolveParams)
-          .countDocuments()
-          .exec();
+ resolveParams.query = model.countDocuments();
+ return beforeQueryHelper(resolveParams, model)

```
`await UserModel.countDocuments().where({ name: 'userName2' })` works perfectly (firstly applies where condition then makes count)


### 3) Tests, need more tests )

**Comment by antoniopresto** [comment]
Nice! Thanks for the quick response. I'll take a look at this right now

**Comment by antoniopresto** [comment]
Hi @nodkz! You see any drawbacks to put the model also in resolveParams?

**Comment by nodkz** [comment]
@antoniopresto Yep, put the `model` to `ExtendedResolveParams`. I want to suggest the same solution.

**Comment by antoniopresto** [comment]
implemented more tests :) I think its complete

File: src/resolvers/__tests__/connection-test.js [hunk 1]
```diff
@@ -217,6 +218,63 @@ describe('connection() resolver', () => {
         expect(result.edges[0].node).toBeInstanceOf(UserModel);
         expect(result.edges[1].node).toBeInstanceOf(UserModel);
       });
+
+      it('should call `beforeQuery` method with non-executed `query` as arg', async () => {
+        const mongooseActions = [];
+
+        UserModel.base.set('debug', function debugMongoose(...args) {
```
Comment (nodkz) [inline src/resolvers/__tests__/connection-test.js]: Nice👍 didn't know such spying way

File: src/resolvers/findById.js [hunk 2]
```diff
@@ -30,8 +31,9 @@ export default function findById<TSource: MongooseDocument, TContext>(
 
       if (args._id) {
         resolveParams.query = model.findById(args._id); // eslint-disable-line
+        resolveParams.model = model; // eslint-disable-line
         projectionHelper(resolveParams);
-        return resolveParams.query.exec();
+        return beforeQueryHelper(resolveParams, model);
```
Comment (nodkz) [inline src/resolvers/findById.js]: Forgot to remove second arg (model)?
All resolvers affected.

**Comment by nodkz** [comment]
Great work! 👍
After removing model I'm ready to merge. But can do it only tomorrow (in 9 hours).

PS. Any example in readme will be very appreciated but not obligatory.

File: src/resolvers/findById.js (hunk 2 above)
Comment (antoniopresto) [reply src/resolvers/findById.js]: nice catch :)

File: src/resolvers/__tests__/connection-test.js (hunk 1 above)
Comment (antoniopresto) [reply src/resolvers/__tests__/connection-test.js]: discovered this week haha

**Comment by antoniopresto** [comment]
Thanks! Your work is inspiring. I can try to help with the readme, but my english doesn't help much hehe

**Comment by nodkz** [comment]
Very cool PR! Thanks! 🙏

**Pull request merged**